ecommerce/extensions/fulfillment/modules.py

```
@@ -103,6 +103,57 @@ def revoke_line(self, line):
         raise NotImplementedError("Revoke method not implemented!")
 
 
+class DoNothingDonationFulfillmentModule(BaseFulfillmentModule):
+    """
+    Fulfillment module for fulfilling donations as a part of LEARNER-2842 - Test Donations on Checkout.
+    If that test, or a follow up test around donations at checkout is not implemented, this module will be reverted.
+    """
+    def supports_line(self, line):
+        """
+        Returns True if the given Line has a donation product.
+        """
+        return line.product.get_product_class().name == "Donation"
+
+    def get_supported_lines(self, lines):
+        """ Return a list of supported lines (that contain a donation product)
+
+        Args:
+            lines (List of Lines): Order Lines, associated with purchased products in an Order.
+
+        Returns:
+            A supported list of lines, unmodified.
+        """
+        return [line for line in lines if self.supports_line(line)]
+
+    def fulfill_product(self, order, lines):
+        """ Fulfills the specified lines in the order.
+        Marks the line status as complete. Does not change anything else.
+
+        Args:
+            order (Order): The Order associated with the lines to be fulfilled
+            lines (List of Lines): Order Lines, associated with purchased products in an Order.
+
+        Returns:
+            The original set of lines, with new statuses set based on the success or failure of fulfillment.
+
+        """
+        for line in lines:
+            line.set_status(LINE.COMPLETE)
+        return order, lines
+
+    def revoke_line(self, line):
+        """ Revokes the specified line.
+        (Not implemented since revoking a noop is meaningless)
+
+        Args:
+            line (Line): Order Line to be revoked.
+
+        Returns:
+            True, if the product is revoked; otherwise, False.
+        """
+        raise NotImplementedError("Revoke method not implemented!")
```

Revocation may be a no-op, but that doesn't mean you can raise an exception. If someone wants a refund, this line will blow up. Just return `True`.

---

why do CouponFulfillmentModule/EnrollmentCodeFulfillmentModule raise a not implemented error?

---

Those products are not refundable.

---

Right, product did not say these should be refundable but I will ask them to change requirements